ai: Add a node embedding watcher

hugoShaka opened the pull request:
This PR adds the embedding core logic in auth:
- add Embeddings service and its local implementation
- add Embedding type and proto message
- add nodeEmbeddingCollector tracking nodes
- add NodeEmbeddingWatcher watching for events adn sending them to the
  collector
- add the Embedder interface and its openai implementation

Fixes https://github.com/gravitational/teleport.e/issues/1464

hugoShaka committed:
ai: add embeddings basic support

- add Embeddings service and its local implementation
- add Embedding type and proto message
- add nodeEmbeddingCollector tracking nodes
- add NodeEmbeddingWatcher watching for events adn sending them to the
  collector
- add the Embedder interface and its openai implementation

xacrimon reviewed (commented):
first pass

xacrimon commented on lib/services/local/embeddings_test.go:
```diff
@@ -0,0 +1,178 @@
+package local
```
missing license notice?

xacrimon commented on lib/services/local/embeddings.go:
```diff
@@ -0,0 +1,164 @@
+package local
```
missing license notice?

xacrimon commented on lib/services/embeddings_test.go:
```diff
@@ -0,0 +1,149 @@
+package services_test
```
missing license notice?

xacrimon commented on lib/services/embeddings.go:
```diff
@@ -0,0 +1,312 @@
+package services
```
missing license notice?

jakule commented on lib/services/embeddings.go:
```diff
@@ -0,0 +1,312 @@
+// embedding on all nodes needing embeddings. The embeddings are then inserted
+// into the vector index. This process is ran asynchronously to reduce the load
+// and leverage OpenAI's batch embedding API.
+func (n *nodeEmbeddingCollector) RunIndexation(ctx context.Context) error {
```
If I understand this correctly, if two goroutines run this method, both will read all nodes, then both will create embeddings and update the index with the same values. Shouldn't we just take the exclusive lock at the beginning and double-check if the embedding generation is needed to avoid it?

hugoShaka replied:
The only sensitive part in this is when the full sync runs and the function getResourcesAndUpdateCurrent acquires the lock.

There's a single routine triggering the embeddings so we should not be doing double work, but even if this happens this does not affect consistency because of the timestamp check.

jakule commented on lib/services/embeddings_test.go:
```diff
@@ -0,0 +1,149 @@
+	return result, nil
+}
+
+func TestNodeEmbeddingWatcher(t *testing.T) {
```
This test tests a lot of stuff. Can we split it into a few smaller/simpler tests? Otherwise, if something is broken, you don't have clear information what exactly happened.

hugoShaka commented on lib/ai/embedding.go:
```diff
@@ -0,0 +1,114 @@
+// ComputeEmbeddings taxes a map of nodes and calls openAI to generate
+// embeddings for those nodes. ComputeEmbeddings is responsible for
+// implementing a retry mechanism if the embedding computation is flaky.
+func (client *Client) ComputeEmbeddings(ctx context.Context, input []string) ([][]float32, error) {
```
I introduced two types: `Vector32` and `Vector64`. I only use Vector32 everywhere, but @jakule's vector index uses float64, so the second type will become handy and we'll be able to write conversion methods.

xacrimon replied:
👍

hugoShaka commented on api/proto/teleport/embedding/v1/embedding.proto:
```diff
@@ -0,0 +1,40 @@
+  types.Metadata metadata = 1;
+
+  // Version is the embedding resource version.
+  string version = 2;
```
Currently, we don't even need a version, but we would if we decide to change the embedding resource. I made this a string to be consistent with the `Version` field of our other resources.

xacrimon replied:
to me, keeping consistent there sounds good

hugoShaka replied:
I dropped the version and store directly proto messages in the backend

hugoShaka commented on lib/services/embeddings.go:
```diff
@@ -0,0 +1,312 @@
+// It keeps tracks of which node has been embedded and which node requires embedding.
+// The embedding happens asynchronously as calling the openAI API by batch is
+// much quicker, stable and efficient.
+type nodeEmbeddingCollector struct {
```
I don't think, 3 of them are here in all other collectors. The initialization channel and its fuse are here to please the resourceCollector interface. Same for the `stale` boolean.

The only synchronization structure used by our custom logic is the mutex.

hugoShaka commented on lib/services/embeddings.go:
```diff
@@ -0,0 +1,312 @@
+	currentNodes map[string]*embeddedNode
+
+	rw    sync.RWMutex
+	stale atomic.Bool
```
I added some comments and revamped the locking mechanism, can you check if this makes more sense and if there's information lacking?

hugoShaka committed:
ai: adapt embeddings to the vector index

hugoShaka committed:
Merge branch 'master' into hugo/ai-embedding-watcher

hugoShaka committed:
fixup! ai: adapt embeddings to the vector index

hugoShaka committed:
fixup! fixup! ai: adapt embeddings to the vector index

jakule commented on lib/service/service.go:
```diff
@@ -1694,6 +1701,27 @@ func (process *TeleportProcess) initAuthService() error {
 	authServer.SetLockWatcher(lockWatcher)
 
+	if authServer.Embeddings != nil {
+		log.Infof("Starting embedding watcher")
```
```suggestion
		log.Debugf("Starting embedding watcher")
```
Let's try not to log in `Info` without a reason.

jakule reviewed (approved):
I'm fine with the code as long as @justinas confirms the OpenAI API key logic.

hugoShaka commented on lib/config/configuration.go:
```diff
@@ -697,6 +697,16 @@ func applyAuthConfig(fc *FileConfig, cfg *servicecfg.Config) error {
 		cfg.Auth.Preference.SetDisconnectExpiredCert(fc.Auth.DisconnectExpiredCert.Value)
 	}
 
+	if fc.Auth.Assist != nil && fc.Auth.Assist.OpenAI != nil {
```
I think the auth endpoint to query embeddings should return an error, and the proxies should handle it gracefully by bailing out and returning an error message saying embeddings must be enabled. I don't think it adds much value to handle cases where embeddings are disabled on the auth side, users will just have a poor UX.

hugoShaka committed:
Update lib/service/service.go

Co-authored-by: Jakub Nyckowski <jakub.nyckowski@goteleport.com>

jakule commented on lib/services/embeddings.go:
```diff
@@ -0,0 +1,396 @@
+// we use in utils.FastMarshall() but with more precision on floats. This
+// is required to reduce the lost precision when storing embeddings in the
+// backend.
+var PreciseJSONConfig = jsoniter.Config{
+	EscapeHTML:                    false,
+	MarshalFloatWith6Digits:       false, // keep precision
+	ObjectFieldMustBeSimpleString: true,  // do not unescape object field
+	SortMapKeys:                   true,
+}.Froze()
+
```
Sorry, I missed that earlier. Can we use the standard JSON library? We had many issues with this one in the past.

justinas commented on lib/service/servicecfg/auth.go:
```diff
@@ -105,6 +105,10 @@ type AuthConfig struct {
+
+	// AssistAPIKey is the OpenAI API key.
+	// TODO: This key will be moved to a plugin once support for plugins is implemented.
+	AssistAPIKey string
```
This isn't persisted into a resource etc. as far as I can tell? Should be fine.

espadolini commented on lib/ai/embedding.go:
```diff
@@ -0,0 +1,165 @@
+// Sha256Hash is the hash of the embedded content. This hash allows to detect if
+// the embedding is still up-to-date or if the content changed and the resource
+// must be re-embedded.
+type Sha256Hash = [32]byte
```
drive-by nitpick

```suggestion
type Sha256Hash = [sha256.Size]byte
```

from `crypto/sha256`

xacrimon commented on lib/ai/embedding.go:
```diff
@@ -0,0 +1,165 @@
+
+const (
+	maxOpenAIEmbeddingsPerRequest = 1000
+	EmbeddingPeriod               = 5 * time.Minute
```
godoc on a public item?

xacrimon commented on lib/services/embeddings.go:
```diff
@@ -0,0 +1,396 @@
+// we use in utils.FastMarshall() but with more precision on floats. This
+// is required to reduce the lost precision when storing embeddings in the
+// backend.
+var PreciseJSONConfig = jsoniter.Config{
```
will be be swapping jsoniter?

jakule commented on lib/ai/embedding.go:
```diff
@@ -0,0 +1,165 @@
+func NewEmbedding(kind, id string, vector Vector64, hash Sha256Hash) *Embedding {
+	return &Embedding{
+		Embedding: &embeddingpb.Embedding{
+			Metadata: &types.Metadata{
+				Name: kind + "/" + id,
+			},
```
I was thinking about when I was working with this code. Is there a reason to keep the metadata here instead of just kind and id? Metadata keeps more information such as namespace, labels, etc. That information is never populated, and I find myself getting the empty value multiple times.

espadolini commented on lib/ai/embedding.go:
```diff
@@ -0,0 +1,165 @@
+// Teleport Assist embeds resources to perform semantic search.
+// The Embedding is named after the embedded resource id and kind. For example
+// the SSH node "bastion-01" has the embedding "node/bastion-01".
+type Embedding struct {
+	*embeddingpb.Embedding
+}
```
Maybe?

```suggestion
type Embedding *embeddingpb.Embedding
```

espadolini commented on lib/services/local/embeddings.go:
```diff
@@ -0,0 +1,178 @@
+// already has an up-to-date embedding stored in the backend, and computes
+// a new embedding otherwise. The newly computed embedding is stored in
+// the backend.
+func (e EmbeddingsService) Embed(ctx context.Context, kind string, resources map[string][]byte) ([]*ai.Embedding, error) {
```
`lib/services/local` doesn't feel like the correct place for this - so far the services have only had to deal with CPU-bound marshaling and unmarshaling and interacting with the backend, not call to some external service.

hugoShaka replied:
I moved the logic out of the Embeddings service. It only performs backend queries and marshalling operations.

espadolini commented on lib/services/local/embeddings.go:
```diff
@@ -0,0 +1,178 @@
+// GetEmbeddings returns all embeddings for a given kind.
+func (e EmbeddingsService) GetEmbeddings(ctx context.Context, kind string) ([]*ai.Embedding, error) {
+	startKey := backend.Key(embeddingsPrefix, kind)
+	result, err := e.GetRange(ctx, startKey, backend.RangeEnd(startKey), backend.NoLimit)
```
This will OOM the auth in a large enough cluster when asking for embeddings for `node`, since we intend to have (at least!) one embedding per active node.

hugoShaka replied:
I switched to a stream. This function is currently called only in tests so it did not required large logic changes. If/when we implement garbage collection we'll be able to use the stream in a safe way.

hugoShaka committed:
address feedback pt.1

hugoShaka committed:
address feedback pt.2: store protobuf message in backend

hugoShaka committed:
address feedback pt.3: have GetEmbeddings return a stream

hugoShaka committed:
Update lib/services/embeddings.go

Co-authored-by: rosstimothy <39066650+rosstimothy@users.noreply.github.com>

espadolini commented on lib/services/local/embeddings.go:
```diff
@@ -52,26 +53,17 @@ func (e EmbeddingsService) GetEmbedding(ctx context.Context, kind, resourceID st
-func (e EmbeddingsService) GetEmbeddings(ctx context.Context, kind string) ([]*ai.Embedding, error) {
+// GetEmbeddings returns a stream of embeddings for a given kind.
+func (e EmbeddingsService) GetEmbeddings(ctx context.Context, kind string) stream.Stream[*ai.Embedding] {
 	startKey := backend.Key(embeddingsPrefix, kind)
```
```suggestion
	startKey := backend.ExactKey(embeddingsPrefix, kind)
```

espadolini commented on lib/services/embeddings.go:
```diff
@@ -0,0 +1,383 @@
+	Embed(ctx context.Context, kind string, resources map[string][]byte) ([]*ai.Embedding, error)
+}
+
+// MarshalEmbedding marshals the ai.Embedding resource to JSON.
```
```suggestion
// MarshalEmbedding marshals the ai.Embedding resource to binary ProtoBuf.
```

espadolini commented on lib/services/embeddings.go:
```diff
@@ -0,0 +1,383 @@
+	return data, nil
+}
+
+// UnmarshalEmbedding unmarshals JSON into an ai.Embedding resource.
```
```suggestion
// UnmarshalEmbedding unmarshals binary ProtoBuf into an ai.Embedding resource.
```

hugoShaka committed:
Apply suggestions from code review

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

hugoShaka committed:
address feedback pt.4: extract embedding logic out of Embeddings service

hugoShaka committed:
fixup! address feedback pt.4: extract embedding logic out of Embeddings service

hugoShaka commented on lib/services/embeddings.go:
```diff
@@ -0,0 +1,396 @@
+
+// NodeCount returns how many nodes are tracked by the collector and if they
+// need embedding.
+func (n *nodeEmbeddingCollector) NodeCount(needsEmbedding bool) int {
+	count := 0
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+	for _, node := range n.currentNodes {
+		if node.needsEmbedding == needsEmbedding {
+			count += 1
+		}
+	}
+	return count
+}
```
We can optimize this function, but I think it would defeat this counter's purpose. Tests need a way to peak into the collector's state to know what is happening (and the tests are living in a separate `services_test` package, a private function is not usable here). Making this counter smarter and more efficient increases the risks of it returning an inaccurate value because of some collector bug.

hugoShaka committed:
address feedback pt.5: simpler error handling when embedding fails

hugoShaka committed:
fix tests pt.1

hugoShaka committed:
fix tests pt.2

hugoShaka committed:
fix tests pt.3

hugoShaka commented on lib/services/embeddings.go:
```diff
@@ -0,0 +1,396 @@
+// It keeps tracks of which node has been embedded and which node requires embedding.
+// The embedding happens asynchronously as calling the openAI API by batch is
+// much quicker, stable and efficient.
+type nodeEmbeddingCollector struct {
```
Is this comment a blocker, or can we address it later when we'll extend embeddings? This PR is somewhat important to unlock the rest of the Assist PRs for the next features and it would be easier for us to redesign this later.

I can extract the embedding logic out of the collector, but this will complexify even more an [already complex locking situation](https://github.com/gravitational/teleport/pull/27204#discussion_r1223328498) as the collector and something else would need to sync to ensure embedding consistency. Do you think of a specific design that would allow consistent asynchronous embedding not triggered in the watcher?

jakule committed:
[Assist] Replace embedding watcher (#27953)

Change the way how the embeddings are calculated. Instead of creating a watcher in Auth, we will process all nodes every hour and process embeddings if any embeddings are missing or any node has been updated.

r0mant reviewed (approved):
Bot.
